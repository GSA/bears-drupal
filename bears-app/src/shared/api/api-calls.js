/**
 * an async fetch to get life-event data.
 * @function
 * @param {string} lifeEvent - The inherited class from
 * @return {JSON || Sring} returns JSON data if succesfull
 */
export async function GETLifeEvent(lifeEvent) {
  // get life-event from location
  if (lifeEvent === undefined) {
    const location = window.location.pathname
    lifeEvent = location.substring(location.lastIndexOf('/') + 1)
  }
  const response = await fetch(`/bears/api/life-event/${lifeEvent}`)
    .then(response => {
      if (response?.ok) {
        return response.json()
      }
      throw new Error(response?.status)
    })
    .then(responseJson => {
      return responseJson?.data.length > 0
        ? responseJson
        : 'Something went wrong.'
    })
    .catch(error => {
      // eslint-disable-next-line no-console
      console.log(error)
      return 'Something went wrong.'
    })
  return response
}
