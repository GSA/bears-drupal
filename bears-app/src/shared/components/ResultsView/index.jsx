import PropTypes from 'prop-types'
import { BenefitAccordionGroup, Heading, StepBackLink } from '../index'

/**
 * a functional component that renders a view of the form benefit state values
 * @component
 * @param {function} handlStepBack inherited ui translations
 * @param {object} ui inherited ui translations
 * @param {array} data inherited benefits data
 * @param {func} setBenefitsArray inherited state handler
 * @param {array} stepDataArray inherited state of inupt values from form entry
 * @return {html} returns a view page of filtered benefits
 */
const ResultsView = ({
  handleStepBack,
  ui,
  data,
  // setBenefitsArray,
  // stepDataArray,
}) => {
  // console.log(data)
  const { heading, stepBackLink } = ui

  // collect all the criteria keys and selected criteria values into an array
  // const selectedCriteria = stepDataArray
  //   .map((item, i) =>
  //     item.section.fieldsets.map(item => {
  //       return {
  //         id: item.fieldset.inputs[0].inputCriteria.id,
  //         values: item.fieldset.inputs[0].inputCriteria.values.find(
  //           value => value.selected
  //         ),
  //       }
  //     })
  //   )
  //   .flat() // we flatten all to have only one array

  // Total Criteria = y
  // Met Criteria = x
  // Not Met Criteria = z

  // Criteria Met Length	Label
  // x === y	"Likely Eligible"
  // z === 0 && x === undefined length > 0 "More Information Needed"
  // Criteria Not Met	Label
  // z > 0	"Not Eligible"

  // collect "Likely Eligible"

  // for each benefit eligiblity compare compare selectedCriteria
  // if there is a criteriakey match in a benefit
  // check that the value === acceptable values

  // function likelyEligible(selectedCriteria, data) {
  //   console.log(selectedCriteria)
  //   data.forEach((item, i) => {
  //     console.log(item.benefit.eligibility)
  //   })
  // }

  // console.log(likelyEligible(selectedCriteria, data))

  // compare the selected criteria array with benefits
  return (
    <div>
      <Heading headingLevel={2}>{heading}</Heading>
      {/* map all the benefits into cards */}
      <BenefitAccordionGroup data={data} entryKey={'benefit'} expandAll />
      {/* {data.map((item, i) => {
        return console.log(item)
      })} */}
      <StepBackLink setCurrent={handleStepBack}>{stepBackLink}</StepBackLink>
    </div>
  )
}

ResultsView.propTypes = {
  handleStepBack: PropTypes.func,
  ui: PropTypes.object,
  data: PropTypes.array,
}

export default ResultsView
