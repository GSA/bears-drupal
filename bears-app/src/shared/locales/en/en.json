{
  "intro": {
    "heading": "Steps",
    "timeIndicator": "Time to complete",
    "steps": {
      "list": [
        {
          "title": "Answer questions"
        },
        {
          "title": "Review Results"
        },
        {
          "title": "Find out how to apply"
        }
      ]
    },
    "notices": {
      "heading": "Before you start",
      "list": [
        {
          "notice": "<div>Your information<strong>will not</strong> be shared, saved, or submitted.</div>"
        },
        {
          "notice": "Provide complete and accurate information to get the most relevant results."
        },
        {
          "notice": "<div><strong>This form is not an application.</strong> You will need to apply for benefits with each agency.</div>"
        }
      ]
    },
    "button": "Start"
  },
  "stepIndicator": {
    "stepBackLink": "Back"
  },
  "requiredLabel": {
    "value": "Required"
  },
  "buttonGroup": [
    {
      "value": "Back"
    },
    {
      "value": "Continue"
    }
  ],
  "reviewSelectionModal": {
    "heading": "Select an option:",
    "buttonGroup": [
      {
        "value": "Review selections"
      },
      {
        "value": "See results"
      }
    ]
  },
  "verifySelectionsView": {
    "heading": "Review and Confirm",
    "buttonGroup": [
      {
        "value": "Back"
      },
      {
        "value": "Confirm"
      }
    ]
  },
  "resultsView": {
    "heading": "Your potential benefits",
    "description": "<div>You must verify your eligibility and visit each agency to apply. <strong>Check with your city, state, or employer for other benefits.</strong></div>",
    "stepBackLink": "Back",
    "eligibleResults": {
      "heading": "Results",
      "description": "You may be eligible for these benefits."
    },
    "notEligibleResults": {
      "heading": "Benefits you did not qualify for",
      "description": "<div>Based on your answers you are not eligible for these benefits. <strong>You may become eligible if you enter additional information or your situation changes.</strong></div>",
      "cta": "See benefits you did not qualify for"
    }
  },
  "notEligibleResults": {
    "heading": "Benefits you did not qualify for",
    "description": "<div>According to your answers you <strong>are not</strong> eligible for these benefits.</div>",
    "heading2": "Non-eligible Results",
    "description2": "Check with your state, city, or employer for other benefits."
  },
  "shareResults": {
    "heading": "Share results",
    "description": "Your information will be viewable. Only share with those you trust.",
    "buttonGroup": [
      {
        "value": "Share link"
      },
      {
        "value": "Email results"
      }
    ]
  },
  "benefitAccordionGroup": {
    "closedState": "Open all",
    "openState": "Close all"
  }
}
