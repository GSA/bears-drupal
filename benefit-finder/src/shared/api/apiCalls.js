/**
 *
 * UTILS FUNCTIONS
 *
 */

/**
 * a function that breaks down a conditional string and compares to a date epoch
 * @function
 * @param {string} conditional - one of >,>=,<,<=,= and a time value
 * @param {object} selectedValue - date object
 * @return {boolean} true or false based on the conditional
 */
export const DateEligibility = ({ selectedValue, conditional }) => {
  // console.log({ selectedValue, conditional })
  // date values
  // "<01-01-1978"
  // "<2years (the deceased died within the last two years)"
  // "<18years"
  // ">=62years"
  // ">=60years"
  // ">=50years"
  // ">18years"

  // get the conditional
  const text = conditional
  const operators = /['>', '>=', '<', '<=', '=']/g
  const operator = text.match(operators)
  const integer = text.match(/\d+/)[0]

  // calculate back
  // get current date
  // subtract integer
  // if a date comes back in date format
  const pattern = /-/
  const conditionalDate = pattern.test(text)
    ? new window.Date(text)
    : new window.Date(
        new Date().getFullYear() - integer,
        new Date().getMonth(),
        new Date().getDate()
      )

  // example selected value for date
  // const value = {
  //   year: 2022,
  //   month: 2,(month index)
  //   day: 2,
  // }

  // calculate selected
  const selectedDate = new window.Date(
    Date.UTC(
      selectedValue.year,
      selectedValue.month,
      selectedValue.day,
      0,
      0,
      0
    )
  )

  const isDateEligible = (operator, conditionalDate, selectedDate) => {
    // ['>', '>=', '<', '<=', '=']
    // epoch time measures the number of seconds that have elapsed since the start of the Unix epoch on January 1st, 1970, at midnight UTC/GMT, minus the leap seconds.

    const x = selectedDate
    const y = new window.Date(
      Date.UTC(
        conditionalDate.getUTCFullYear(),
        conditionalDate.getUTCMonth(),
        conditionalDate.getUTCDate(),
        0,
        0,
        0
      )
    )

    const diff = !pattern.test(text)
      ? y.getTime() - x.getTime()
      : y.getTime() - x.getTime()

    switch (operator.length && operator.join('')) {
      case '>':
        return diff > 0
      case '>=':
        return diff >= 0
      case '<':
        return diff < 0
      case '<=':
        return diff <= 0
      case '=':
        return +diff === 0
      default:
        return false
    }
  }
  return isDateEligible(operator, conditionalDate, selectedDate)
}

/**
 * a function that searches an array for element or child element match to a key
 * @function
 * @param {array} arr - array of fieldsets
 * @param {string} criteriaKey - unique identifier
 * @return {object} returns the element if found
 */
export async function FindCriteria(arr, criteriaKey) {
  let element

  element = arr.find(element => element.fieldset.criteriaKey === criteriaKey)

  // if the element returns as undefined, then go check for children
  if (element === undefined) {
    arr.forEach(item => {
      item.fieldset.children.forEach(childElement => {
        if (childElement.fieldsets[0].fieldset.criteriaKey === criteriaKey) {
          element = childElement.fieldsets[0]
        }
      })
    })
  }
  return element
}

/**
 * a function that takes the search string and parses them into an array
 * @function
 * @param {string} url - array of fieldsets
 * @return {array} returns array of objects
 */
export function GetQueryParams(url) {
  const paramArray = url.slice(url.indexOf('?') + 1).split('&')
  const selectedArray = paramArray.map(elem => {
    const [key, val] = elem.split('=')

    return {
      criteriaKey: key,
      value: decodeURIComponent(val),
    }
  })
  return selectedArray
}

/**
 *
 * GET FUNCTIONS
 *
 */

/**
 * a function that determines our langage state based on  the prefix of our URL
 * based on a string match in the pathname of the window object
 * @function
 */
export const Language = () => {
  const string = /^\/es/
  return string.test(window.location.pathname) ? 'es' : 'en'
}

/**
 * an async fetch to get life-event data.
 * @function
 * @param {string} lifeEvent - The inherited class from
 * @return {JSON} returns JSON data if succesfull
 */
export async function LifeEvent(lifeEvent) {
  let language, params, mode
  // get life-event from location
  if (lifeEvent === undefined) {
    const string = /^\/es/
    language = string.test(window.location.pathname) ? 'es_' : ''
    const windowQuery = window.location.search
    params = new URLSearchParams(windowQuery)
    mode = params.get('mode') === 'draft' ? `${params.get('mode')}/` : ''
    const location = window.location.pathname
    lifeEvent = location.substring(location.lastIndexOf('/') + 1)
  }
  const response = await fetch(
    `/sites/default/files/benefit-finder/api/${mode}life-event/${language}${lifeEvent}.json`
  )
    .then(response => {
      if (response?.ok) {
        return response.json()
      }
      throw new Error(response?.status)
    })
    .then(responseJson => {
      return responseJson?.data ? responseJson : 'Something went wrong.'
    })
    .catch(error => {
      // eslint-disable-next-line no-console
      console.log(error)
      return 'Something went wrong.'
    })
  return response
}

/**
 * a function to get the value object of selected fieldset values
 * @function
 * @param {object} item - fieldset object
 * @return {object} returns the object with a selected key
 */
export const SelectedValue = item =>
  item &&
  item.fieldset.inputs[0].inputCriteria.values.find(value => value.selected)

/**
 * a function to get array of children
 * @function
 * @param {object} item - parent fieldset object
 * @return {array} returns an array of children related to the param
 */
export const Children = item =>
  item && item.fieldset?.children.map(childItem => childItem.fieldsets[0])

/**
 * a function to get an array of all selected values
 * @function
 * @param {array} data - array of fieldset data
 * @return {array} returns the selected values
 */
export const SelectedValueAll = data =>
  data &&
  data
    .flatMap((item, i) =>
      item.section.fieldsets.flatMap(item => {
        const selected = GET.SelectedValue(item)
        return (
          selected && [
            {
              criteriaKey: item.fieldset.inputs[0].inputCriteria.id,
              values: selected,
            },
            GET.Children(item) &&
              GET.Children(item)
                .flatMap(
                  childItem =>
                    GET.SelectedValue(childItem) && {
                      criteriaKey: childItem.fieldset.criteriaKey,
                      values: GET.SelectedValue(childItem),
                    }
                )
                .filter(element => element !== undefined),
          ]
        )
      })
    )
    .flat() // we flatten all to have only one array
    .filter(element => element !== undefined) // remove undefined

/**
 * a function that filters determines eligibility of benefits by selected values
 * @function
 * @param {array} selectedCriteria - array of selected fieldset values
 * @param {array} data - array of benefits
 * @return {array} returns the selected values
 */
export const ElegibilityByCriteria = (selectedCriteria, data) => {
  // return all eligible items
  const eligibleItems =
    data &&
    data.map(item => {
      // find all eligibility items that are matches to criteria key
      selectedCriteria.forEach(selected => {
        // match item to criteria key
        const criteriaEligibility = item.benefit.eligibility.find(
          criteria => criteria.criteriaKey === selected.criteriaKey
        )

        // determine it's eligiblity
        if (criteriaEligibility !== undefined) {
          // look for eligible matches
          const isEligible = () => {
            let eligibility

            if (typeof selected.values.value === 'object') {
              eligibility = criteriaEligibility.acceptableValues.find(value =>
                UTILS.DateEligibility({
                  selectedValue: selected.values.value,
                  conditional: value,
                })
              )
            } else {
              eligibility = criteriaEligibility.acceptableValues.find(
                value => value === selected.values.value
              )
            }
            return eligibility
          }

          // undefined === false
          criteriaEligibility.isEligible = !!isEligible()
        }
      })
      return item
    })
  // merge all arrays and objects into one array
  const mergedEligibleItems = eligibleItems && [].concat(...eligibleItems)
  return mergedEligibleItems
}

/**
 *
 * PUT FUNCTIONS
 *
 */

/**
 * an async fetch to get life-event data.
 * @function
 * @param {string} criteriaKey
 * @param {object} currentData
 * @param {function} setCurrentData
 * @param {string} eventTargetValue
 * @return {JSON || String} returns JSON data if succesfull
 */
export async function Data(
  criteriaKey,
  currentData,
  setCurrentData,
  eventTargetValue
) {
  // copy current data state
  const newData = { ...currentData }

  UTILS.FindCriteria(newData.section.fieldsets, criteriaKey)
    .then(foundCriteria => {
      if (foundCriteria) {
        const inputValues =
          foundCriteria.fieldset?.inputs[0].inputCriteria.values

        inputValues.forEach(value => {
          if (value.value === eventTargetValue) {
            value.selected = true
          } else {
            delete value.selected
          }
        })
        return setCurrentData(newData)
      }
    })
    .catch(error => {
      // eslint-disable-next-line no-console
      console.log(error)
      return 'Something went wrong.'
    })
}

/**
 * an async fetch to get life-event data.
 * @function
 * @param {string} criteriaKey
 * @param {object} currentData
 * @param {function} setCurrentData
 * @param {string} eventTargetValue
 * @param {string} eventTargetID
 * @return {JSON || String} returns JSON data if succesfull
 */
export async function DataDate(
  criteriaKey,
  currentData,
  setCurrentData,
  eventTargetValue,
  eventTargetID
) {
  // copy current data state
  const newData = { ...currentData }

  UTILS.FindCriteria(newData.section.fieldsets, criteriaKey)
    .then(foundCriteria => {
      if (foundCriteria) {
        const inputValues =
          foundCriteria.fieldset?.inputs[0].inputCriteria.values

        if (eventTargetID) {
          if (eventTargetID.includes('day')) {
            inputValues[0].value.day = eventTargetValue
          }
          if (eventTargetID.includes('month')) {
            inputValues[0].value.month = eventTargetValue
          }
          if (eventTargetID.includes('year')) {
            inputValues[0].value.year = eventTargetValue
          }

          inputValues[0].value = { ...inputValues[0].value }
        } else {
          inputValues[0].value = eventTargetValue
        }
        inputValues[0].selected = true
        return setCurrentData(newData)
      }
    })
    .catch(error => {
      // eslint-disable-next-line no-console
      console.log(error)
      return 'Something went wrong.'
    })
}

/**
 * updates our data based on the window.location.search value
 * @function
 * @param {string} windowQuery
 * @param {array} stepDataArray
 * @param {function} setCurrentData
 * @param {string} sharedToken
 */
export const DataFromParams = (
  windowQuery,
  stepDataArray,
  setCurrentData,
  sharedToken
) => {
  const params = UTILS.GetQueryParams(decodeURI(windowQuery))
  params.filter(param => param.criteriaKey !== sharedToken)

  stepDataArray.forEach(arr => {
    arr.completed = true
    params.forEach(param => {
      const v = param.value.includes('{')
        ? JSON.parse(param.value)
        : param.value
      v !== undefined && typeof v === 'object'
        ? PUT.DataDate(param.criteriaKey, arr, setCurrentData, v)
        : PUT.Data(param.criteriaKey, arr, setCurrentData, v)
    })
  })
}

export const GET = {
  Children,
  ElegibilityByCriteria,
  LifeEvent,
  Language,
  SelectedValue,
  SelectedValueAll,
}

export const PUT = {
  Data,
  DataDate,
  DataFromParams,
}

export const UTILS = {
  FindCriteria,
  DateEligibility,
  GetQueryParams,
}
