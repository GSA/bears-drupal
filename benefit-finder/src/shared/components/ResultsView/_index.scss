@use '../../styles/colors/_index.scss' as color;
@use '../../styles/breakpoints/_index.scss' as *;
@use '../../styles/space/_index.scss' as space;

#benefit-finder {
  .result-view-details {
    padding: space.$space-lg;

    button.step-back-link {
      color: color.$medium-grey !important;
      background-color: color.$white !important;
      width: auto !important;
    }
  }

  .result-view-benefits,
  .result-view-unmet,
  .result-view-relvant-benefits,
  .result-view-share-results {
    padding: space.$space-md 0;
    border-bottom: 2px solid color.$base-light-grey;
  }

  .result-view-unmet {
    button {
      margin-bottom: space.$space-md;
    }

    p {
      padding-bottom: space.$padding-bottom-md;
    }
  }

  .result-view-share-results {
    text-align: center;
  }

  .result-view-share-results-button-group {
    button {
      margin-bottom: space.$space-sm;
    }
  }

  @media (width >= $desktop) {
    .result-view-details {
      padding: space.$space-md-plus 164px;
    }
  }
}
