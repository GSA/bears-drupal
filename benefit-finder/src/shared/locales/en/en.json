{
  "intro": {
    "heading": "Steps",
    "timeIndicator": "Time to complete",
    "steps": {
      "list": [
        {
          "title": "Answer questions"
        },
        {
          "title": "Review Results"
        },
        {
          "title": "Find out how to apply"
        }
      ]
    },
    "notices": {
      "heading": "Before you start:",
      "list": [
        {
          "notice": "<div>Your information<strong> will not</strong> be shared, saved, or submitted.</div>"
        },
        {
          "notice": "Provide complete and accurate information to get the most relevant results."
        },
        {
          "notice": "<div><strong>This form is not an application.</strong> You will need to apply for benefits with each agency.</div>"
        }
      ]
    },
    "button": "Start"
  },
  "stepIndicator": {
    "stepBackLink": "Back"
  },
  "requiredLabel": {
    "value": "Required"
  },
  "alertBanner": {
    "heading": "Error status",
    "description": "All fields marked required must be completed."
  },
  "buttonGroup": [
    {
      "value": "Back"
    },
    {
      "value": "Continue"
    }
  ],
  "reviewSelectionModal": {
    "heading": "Select an option:",
    "buttonGroup": [
      {
        "value": "Review selections"
      },
      {
        "value": "See results"
      }
    ]
  },
  "verifySelectionsView": {
    "heading": "Review and Confirm",
    "buttonGroup": [
      {
        "value": "Back"
      },
      {
        "value": "Confirm"
      }
    ]
  },
  "resultsView": {
    "qualified": {
      "chevron": {
        "heading": "Your Custom List",
        "description": "<div><p><strong>Here is your list of potential Federal Benefits.</strong> Your state, city or employer may offer other benefits. Visit each agency providing the benefit to learn if you are truly eligible and conduct an actual application.</p></div>"
      },
      "heading": "Results",
      "description": "<div><strong>Based on your answers you meet the criteria to qualify for these benefits:</strong></div>"
    },
    "notQualified": {
      "chevron": {
        "heading": "Benefits You Did Not Qualify For",
        "description": ""
      },
      "heading": "Results",
      "description": "<div><strong>Based on your answers you did not meet the criteria to qualify for these benefits:</strong></div>"
    },
    "stepBackLink": "Back",
    "eligibleResults": {
      "heading": "Results",
      "description": "You may be eligible for these benefits."
    },
    "notEligibleResults": {
      "heading": "Benefits you did not qualify for",
      "description": "<div>Based on your answers you are not eligible for these benefits. <strong>You may become eligible if you enter additional information or your situation changes.</strong></div>",
      "cta": "See benefits you did not qualify for"
    },
    "relativeBenefits": {
      "heading": "Other benefits that might be relevant to you"
    },
    "shareResults": {
      "heading": "Sharing and printing",
      "shareButton": "Share link",
      "emailButton": "Email results",
      "emailSubject": "Results from USAGovâ€™s benefit finder",
      "printButton": "Print results",
      "description": "Copy a link to this page with the criteria you selected above. Make sure to share it only with those you trust as your answers will be visible."
    }
  },
  "shareResults": {
    "heading": "Share results",
    "description": "Your information will be viewable. Only share with those you trust.",
    "buttonGroup": [
      {
        "value": "Share link"
      },
      {
        "value": "Email results"
      }
    ]
  },
  "benefitAccordionGroup": {
    "closedState": "Open all",
    "openState": "Close all"
  },
  "sectionHeadings": {
    "start": "Complete Questions",
    "continue": "Questions Continued",
    "final": "Final Questions"
  },
  "select": {
    "labelSelect": "Select",
    "defaultValue": "Click to Select",
    "dateDefaultValue": "-Select-"
  },
  "date": {
    "alert": "Date should follow the format 01 - January 22 1977",
    "labelDay": "Day",
    "labelMonth": "Month",
    "labelYear": "Year",
    "monthOptions": [
      {
        "label": "01 - January"
      },
      {
        "label": "02 - February"
      },
      {
        "label": "03 - March"
      },
      {
        "label": "04 - April"
      },
      {
        "label": "05 - May"
      },
      {
        "label": "06 - June"
      },
      {
        "label": "07 - July"
      },
      {
        "label": "08 - August"
      },
      {
        "label": "09 - September"
      },
      {
        "label": "10 - October"
      },
      {
        "label": "11 - November"
      },
      {
        "label": "12 - December"
      }
    ]
  }
}
