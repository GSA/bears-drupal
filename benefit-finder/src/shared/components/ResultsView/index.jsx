import { useEffect, useState } from 'react'
import * as apiCalls from '../../api/apiCalls'
import PropTypes from 'prop-types'
import {
  BenefitAccordionGroup,
  Button,
  EmailButton,
  Heading,
  StepBackLink,
  Card,
  Chevron,
  ShareButton,
} from '../index'
import createMarkup from '../../utils/createMarkup'
import './_index.scss'

/**
 * a functional component that renders a view of the form benefit state values
 * @component
 * @param {function} handlStepBack inherited ui translations
 * @param {object} ui inherited ui translations
 * @param {array} data inherited benefits data
 * @param {func} setBenefitsArray inherited state handler
 * @param {array} stepDataArray inherited state of inupt values from form entry
 * @return {html} returns a view page of filtered benefits
 */
const ResultsView = ({
  handleStepBack,
  ui,
  data,
  stepDataArray,
  relevantBenefits,
}) => {
  const {
    stepBackLink,
    notQualified,
    qualified,
    notEligibleResults,
    resultsRelativeBenefits,
    shareResults,
  } = ui

  const [notQualifiedView, setNotQualifiedView] = useState(false)

  const handleViewToggle = () => {
    setNotQualifiedView(!notQualifiedView)
    window.scrollTo(0, 0)
  }

  useEffect(() => {
    window.scrollTo(0, 0)
  }, [])

  // compare the selected criteria array with benefits
  return (
    <div className="result-view" data-testid="result-view">
      <Chevron
        heading={
          notQualifiedView === false
            ? qualified.chevron.heading
            : notQualified.chevron.heading
        }
        description={
          notQualifiedView === false
            ? qualified.chevron.description
            : notQualified.chevron.description
        }
      />
      <div className="grid-container">
        <div className="result-view-details">
          {notQualifiedView === false ? (
            <StepBackLink setCurrent={handleStepBack}>
              {stepBackLink}
            </StepBackLink>
          ) : (
            <Button
              className="step-back-link"
              onClick={() => setNotQualifiedView(false)}
              unstyled
            >
              {stepBackLink}
            </Button>
          )}
          <Heading className="result-view-heading" headingLevel={3}>
            {notQualifiedView ? notQualified.heading : qualified.heading}
          </Heading>
          <div
            className="result-view-description"
            dangerouslySetInnerHTML={
              notQualifiedView
                ? createMarkup(notQualified.description)
                : createMarkup(qualified.description)
            }
          />
          {/* map all the benefits into cards */}
          <div className="result-view-benefits">
            <BenefitAccordionGroup
              data={
                stepDataArray &&
                apiCalls.GET.ElegibilityByCriteria(
                  apiCalls.GET.SelectedValueAll(stepDataArray),
                  data
                )
              }
              entryKey={'benefit'}
              notQualifiedView={notQualifiedView}
              expandAll
              ui={ui}
            />
          </div>
          {notQualifiedView === false && (
            <div className="result-view-unmet">
              <Heading className="result-view-unmet-heading" headingLevel={3}>
                {notEligibleResults?.heading}
              </Heading>
              <p
                dangerouslySetInnerHTML={createMarkup(
                  notEligibleResults?.description
                )}
              />
              <Button onClick={handleViewToggle}>
                {notEligibleResults?.cta}
              </Button>
            </div>
          )}
          {relevantBenefits?.length > 0 && (
            <div className="result-view-relvant-benefits">
              <Heading
                className="result-view-relvant-benefits-heading"
                headingLevel={3}
              >
                {resultsRelativeBenefits?.heading}
              </Heading>
              {relevantBenefits && (
                <ul className="add-list-reset">
                  {relevantBenefits[0] && (
                    <div key="benefit-card-one">
                      <Card
                        className="relative-benefit-card"
                        title={`${relevantBenefits[0].lifeEvent.title}`}
                        body={`${relevantBenefits[0].lifeEvent.body}`}
                        cta={`${relevantBenefits[0].lifeEvent.cta}`}
                        href={`${relevantBenefits[0].lifeEvent.link}`}
                        carrotType={2}
                      ></Card>
                    </div>
                  )}
                  {relevantBenefits[1] && (
                    <div key="benefit-card-two">
                      <Card
                        className="relative-benefit-card"
                        title={`${relevantBenefits[1].lifeEvent.title}`}
                        body={`${relevantBenefits[1].lifeEvent.body}`}
                        cta={`${relevantBenefits[1].lifeEvent.cta}`}
                        href={`${relevantBenefits[1].lifeEvent.link}`}
                        carrotType={2}
                      ></Card>
                    </div>
                  )}
                </ul>
              )}
            </div>
          )}
          <div className="result-view-share-results">
            <Heading
              className="result-view-share-results-heading"
              headingLevel={3}
            >
              {shareResults?.heading}
            </Heading>
            <p>{shareResults?.description}</p>
            <div className="result-view-share-results-button-group">
              <ShareButton
                ui={shareResults}
                data={
                  stepDataArray && apiCalls.GET.SelectedValueAll(stepDataArray)
                }
              />
              <EmailButton
                ui={shareResults}
                data={
                  stepDataArray && apiCalls.GET.SelectedValueAll(stepDataArray)
                }
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  )
}

ResultsView.propTypes = {
  handleStepBack: PropTypes.func,
  ui: PropTypes.object,
  data: PropTypes.array,
}

export default ResultsView
