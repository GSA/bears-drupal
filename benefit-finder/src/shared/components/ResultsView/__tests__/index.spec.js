import * as apiCalls from '../../../api/apiCalls'
import ResultsView from '../index.jsx'
import * as en from '../../../locales/en/en.json'

// import react-testing methods
import { render, screen } from '@testing-library/react'
// add custom jest matchers from jest-dom
import '@testing-library/jest-dom'
import content from '../../../api/mock-data/current.js'
const { data } = JSON.parse(content)

const scenarios = {
  death: [
    {
      scenario: 1,
      windowQuery:
        '?applicant_date_of_birth=%7B"month"%3A"3"%2C"day"%3A"5"%2C"year"%3A"1960"%7D&applicant_relationship_to_the_deceased=Spouse&applicant_marital_status=Widowed&applicant_citizen_status=Yes&applicant_care_for_child=Yes&applicant_paid_funeral_expenses=Yes&deceased_date_of_death=%7B"month"%3A"1"%2C"day"%3A"3"%2C"year"%3A"2022"%7D&deceased_death_location_is_US=Yes&deceased_paid_into_SS=Yes&deceased_public_safety_officer=No&deceased_miner=No&deceased_american_indian=No&deceased_died_of_COVID=Yes&deceased_served_in_active_military=No&shared=true%27',
    },
  ],
}

const windowQuery = scenarios.death[0].windowQuery // Returns:'?q=123'
const benfitsArray = [...data.benefits]
const sharedToken = 'shared'
let stepDataArray
let currentData
let expectedUpdate
// mock useState Function
function setCurrentData(updatedData) {
  currentData = updatedData
}

// handle window.scrollTo
const noop = () => {}
Object.defineProperty(window, 'scrollTo', { value: noop, writable: true })
beforeAll(() => {
  stepDataArray = [...data.lifeEventForm.sectionsEligibilityCriteria]
  setCurrentData(stepDataArray[0])
  apiCalls.PUT.DataFromParams(
    windowQuery,
    stepDataArray,
    setCurrentData,
    sharedToken
  )
  expectedUpdate =
    currentData.section.fieldsets[1].fieldset.inputs[0].inputCriteria.values
})

// render view without data
test('loads view', async () => {
  const view = render(<ResultsView ui={en.resultsView} />)
  await screen.findByTestId('result-view')
  expect(view.baseElement).toMatchSnapshot()
})

// render view with data
test('scenario 1 loads in view', async () => {
  expect(expectedUpdate[0]).toHaveProperty('selected', true)
  const view = render(
    <ResultsView
      ui={en.resultsView}
      stepDataArray={stepDataArray}
      data={benfitsArray}
    />
  )
  await screen.findByTestId('result-view')
  expect(view.baseElement).toMatchSnapshot()

  await screen.findAllByTestId('applicant_relationship_to_the_deceased')
  expect(view.baseElement).toMatchSnapshot()
})
