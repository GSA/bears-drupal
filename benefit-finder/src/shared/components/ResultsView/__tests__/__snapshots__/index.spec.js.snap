// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultsView renders a match to the previous snapshot 1`] = `
<div
  className="result-view"
>
  <div
    className="chevron"
  >
    <div
      className="grid-container"
    >
      <h1
        className=""
        id="skip-to-h1"
      >
        Your Custom List
      </h1>
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": "<div><p><strong>Here is your list of potential Federal Benefits.</strong> Your state, city or employer may offer other benefits.Visit each agency providing the benefit to learn if you are truly eligible and conduct an actual applications.</p></div>",
          }
        }
      />
    </div>
  </div>
  <div
    className="grid-container"
  >
    <div
      className="result-view-details"
    >
      <button
        className=""
        onClick={[Function]}
        type="button"
      >
        Back
      </button>
      <h3
        className=""
        id=""
      >
        Results
      </h3>
      <p
        dangerouslySetInnerHTML={
          Object {
            "__html": "<div><strong>Based on your answers you meet the criteria to qualify for these benefits:</strong></div>",
          }
        }
      />
      <div
        className="result-view-benefits"
      >
        <div
          className="benefit-accordion-group"
        >
          <button
            className=""
            onClick={[Function]}
            type="button"
          >
            Expand all +
          </button>
        </div>
      </div>
      <div
        className="result-view-unmet"
      >
        <h3
          className=""
          id=""
        >
          Benefits you did not qualify for
        </h3>
        <p
          dangerouslySetInnerHTML={
            Object {
              "__html": "<div>Based on your answers you are not eligible for these benefits. <strong>You may become eligible if you enter additional information or your situation changes.</strong></div>",
            }
          }
        />
        <button
          className=""
          onClick={[Function]}
          type="button"
        >
          See benefits you did not qualify for
        </button>
      </div>
      <div
        className="result-view-relvant-benefits"
      >
        <h3
          className=""
          id=""
        >
          Other benefits that might be relevant to you
        </h3>
        <ul
          className="add-list-reset"
        >
          <li>
            <li
              className=""
            >
              <a>
                <div
                  className="usa-card__container"
                >
                  <div
                    className="usa-card__header"
                  >
                    <h3
                      className=""
                      id=""
                    >
                      Approaching Retirement
                    </h3>
                  </div>
                  <p
                    className=""
                  >
                    Learn More
                  </p>
                </div>
              </a>
            </li>
          </li>
          <li>
            <li
              className=""
            >
              <a>
                <div
                  className="usa-card__container"
                >
                  <div
                    className="usa-card__header"
                  >
                    <h3
                      className=""
                      id=""
                    >
                      Living with disability or illness
                    </h3>
                  </div>
                  <p
                    className=""
                  >
                    Learn More
                  </p>
                </div>
              </a>
            </li>
          </li>
        </ul>
      </div>
      <div
        className="result-view-share-results"
      >
        <h3
          className=""
          id=""
        >
          Sharing and printing
        </h3>
        <div
          className="result-view-share-results-button-group"
        >
          <button
            className=""
            onClick={[Function]}
            type="button"
          >
            Share Link
          </button>
          <button
            className=""
            onClick={[Function]}
            type="button"
          >
            Email Results
          </button>
          <button
            className=""
            onClick={[Function]}
            type="button"
          >
            Print Results
          </button>
        </div>
        <p>
          Copy a link to this page with the criteria you selected above. Make sure to share it only with those you trust as your answers will be visible.
        </p>
      </div>
    </div>
  </div>
</div>
`;
